libncftp/samples/ncftpget/ncftpget.1
.TH ncftpget 1 "NcFTP Software" ncftpget
.SH NAME
ncftpget - Internet file transfer program for scripts
.SH "SYNOPSIS"
.PP
.B ncftpget
.RI [ "options" ]
.I "remote-host" "local-directory" "remote-files..."
.PP
.B ncftpget
-f login.cfg
.RI [ "options" ]
.I "local-directory" "remote-files..."
.PP
.B ncftpget
.RI [ "options" ]
.I "ftp://url.style/host/path/name"
.\"-------
.SH "OPTIONS"
.\"-------
.SS
Command line flags:
.TP 8
.RI "-u " "XX"
Use username
.I XX
instead of anonymous.
.TP 8
.RI "-p " "XX"
Use password
.I XX
with the username.
.TP 8
.RI "-P " "XX"
Use port number
.I XX
instead of the default FTP service port (21).
.TP 8
.RI "-d " "XX"
Use the file
.I XX
for debug logging.
.TP 8
-a
Use ASCII transfer type instead of binary.
.TP 8
.RI "-t " "XX"
Timeout after
.I XX
seconds.
.TP 8
-v/-V
Do (do not) use progress meters.
The default is to use progress meters if the output stream is a TTY.
.TP 8
.RI "-f " "XX"
Read the file
.I XX
for host, user, and password information.
.TP 8
-A
Append to local files, instead of overwriting them.
.TP 8
-z/-Z
Do (do not) try to resume transfers.
The default is to try to resume (\-z).
.TP 8
-F
Use passive (PASV) data connections.
.TP 8
-DD
Delete remote file after successfully downloading it.
.TP 8
-R
Recursive mode; copy whole directory trees.
.TP 8
.RI "-r " "XX"
Redial a maximum of 
.I XX
times until connected to the remote FTP server.
.\"-------
.SH "DESCRIPTION"
.\"-------
.PP
The
purpose of
.I ncftpget
is to do file transfers from the command-line
without entering an interactive shell.
This lets you write shell scripts or other unattended
processes that can do FTP.
It is also useful for advanced users who
want to retrieve files from the shell command line without
entering an interactive FTP program such as
.IR ncftp "."
.PP
One particularly useful feature of this program is that
you can give it a
.I uniform resource locator
as the only argument and the program will download
that file.
You can then copy and paste from your web browser
or newsreader and use that URL.
Example:
.RS 4
.sp
$ cd /tmp
.br
$ ncftpget ftp://ftp.ncftp.com/pub/ncftp/ncftp.tar.Z
.br
$ zcat ncftp.tar.Z | tar xf -
.br
.RE
.PP
By default the program tries to open the remote host
and login anonymously, but you can specify a username
and password information.
The
.B -u
option is used to specify the username to login as,
and the
.B -p
option is used to specify the password.
If you are running the program from the shell, you may
omit the
.B -p
option and the program will prompt you for the password.
.PP
Using the 
.B -u
and
.B -p
options are not recommended, because your account information
is exposed to anyone who can see your shell script or your
process information.  For example, someone using the
.I ps
program could see your password while the program runs.
.PP
You may use the
.B -f
option instead to specify a file with the account information.
However, this is still not secure because anyone who
has read access to the information file can see the
account information.
Nevertheless, if you choose to use the
.B -f
option the file should look something like this:
.RS 4
.sp
host Bozo.probe.net
.br
user gleason
.br
pass mypasswd
.br
.sp
.RE
Don't forget to change the permissions on this file
so no one else can read them.
.PP
The
.B -d
option is very useful when you are trying to diagnose
why a file transfer is failing.
It prints out the
entire FTP conversation to the file you specify, so
you can get an idea of what went wrong.  
If you specify the special name
.I stdout
as the name of the debugging output file, the output
will instead print to the screen.  Example:
.RS 4
.sp
$ ncftpget -d stdout ftp.ncftp.com . /pub/ncftp/README
.br
220: ftp.ncftp.com FTP server ready.
.br
Connected to ftp.ncftp.com.
.br
Cmd: USER anonymous
.br
331: Guest login ok, send your complete e-mail address as password.
.br
Cmd: PASS xxxxxxxx
.br
230: Welcome!
.br
Logged in to ftp.ncftp.com as anonymous.
.br
Cmd: TYPE I
.br
200: Type set to I.
.br
Cmd: SIZE /pub/ncftp/README
.br
550: /pub/ncftp/README: not a plain file.
.br
Cmd: MDTM /pub/ncftp/README
.br
550: /pub/ncftp/README: No such file or directory.
.br
Cmd: PORT 192,168,1,17,6,76
.br
200: PORT command successful.
.br
Cmd: RETR /pub/ncftp/README
.br
550: /pub/ncftp/README: No such file OR directory.
.br
ncftpget: file retrieval error: could not start data transfer.
.br
Cmd: QUIT
.br
221: Goodbye.
.br
.RE
.PP
Using ASCII mode is helpful when the text format of your host
differs from that of the remote host.
For example, if you are retrieving a \.TXT file from
a Windows-based host to a UNIX system, you could use the
.B -a
flag which would use ASCII transfer mode so that the file
created on the UNIX system would be in the UNIX text
format instead of the MS\-DOS text format.
.PP
You can retrieve an entire directory tree of files by
using the
.B -R
flag.
However, this will work only if the remote FTP server
is a UNIX server, or emulates UNIX's list output.
Example:
.RS 4
.sp
$ ncftpget -R ftp.ncftp.com /tmp /pub/ncftp
.br
.sp
.RE
This would create a /tmp/ncftp hierarchy.
.\"-------
.SH "DIAGNOSTICS"
.\"-------
.PP
.I ncftpget
returns the following exit values:
.TP 8
0
Success.
.TP 8
1
Could not connect to remote host.
.TP 8
2
Could not connect to remote host - timed out.
.TP 8
3
Transfer failed.
.TP 8
4
Transfer failed - timed out.
.TP 8
5
Directory change failed.
.TP 8
6
Directory change failed - timed out.
.TP 8
7
Malformed URL.
.TP 8
8
Usage error.
.TP 8
9
Error in login configuration file.
.TP 8
10
Library initialization failed.
.TP 8
11
Session initialization failed.
.\"-------
.SH "AUTHOR"
.\"-------
.PP
Mike Gleason, NcFTP Software (http://www.NcFTP.com).
.\"-------
.SH "SEE ALSO"
.\"-------
.PP
.IR ncftpput (1),
.IR ncftp (1),
.IR ftp (1),
.IR rcp (1),
.IR tftp (1).
.PP
.IR "LibNcFTP" " (http://www.ncftp.com/libncftp/)."


doc/CHANGELOG.txt
LibNcFTP Change Log:
===================

3.2.6, 2016-11-12

   + If a recursive download operation is also requested with delete mode,
     attempt to remove empty directories after all files have completed
     successfully.

   + To avoid problems on servers that penalize clients for trying to CWD
     with a non-directory, FTPRmdirRecursive now ensures sub-pathnames
     are directories before trying FTPChdir.

   + No longer trying to utime() after every single block on downloads,
     which could cause noticable performance degradation when the local
     filesystem was not local.

   + Changed behavior of resuming downloads where the timestamp wasn't
     preserved (because of the utime change, above). The new behavior is
     to resume the download when the local copy has a recent timestamp
     (less than a week).

   + You can now disable use of MFMT like you could similarly disable
     SITE UTIME.

   + You can now explicitly set hasAPPE to 0 to prevent it being used
     when hasREST is also 0.

   + The library now caches the most recent value of PWD and uses it
     to prevent redundant PWD commands.

   + Now able to use sendfile() for uploads, on Linux/FreeBSD/Mac.
     Progress reports work too, with a small performance penalty.


3.2.5, 2011-01-17

   + If a server's MLSx output returns both the "UNIX.mode" and the "perm"
     parameters, UNIX.mode will take precedence (Thanks, Jay Kulpinski).


3.2.4, 2010-04-07

   + Compatibility fixes for FreeBSD 8.

   + Support for local validation of DNSSEC when combined with libraries
     available from the DNSSEC Tools project (Thanks, Robert Story).

   + Directory listings now assume a line with no year implies that the
     current year should be used if the date is no more than 2 days into
     the future.  Previously, for UNIX ls compatibility, up to 6 months into
     the future was allowed.

   + Compatibility fixes for MinGW32 (Windows GCC).

   + Microsoft disabled recursive directory listings altogether when fixing
     KB975254 (Thanks, Andrew Coggeshall).  When entire directories are
     downloaded, instead of using "LIST -R" for one efficient listing of
     all files, the directories are manually traversed with one directory
     listing for each subdirectory.  This will improve reliability of
     recursive downloads, except for the case when their are circular
     symbolic links (which is why "LIST -R" had been favored, which the
     server can easily handle).

   + Compatibility fixes for Mac OS X for building from source code
     (Thanks, Mathieu Rene)



3.2.3, 2009-07-24

   + Using MFMT command to set remote timestamps, if available, in
     preference to SITE UTIME or MDTM.

   + New configure flag, --enable-ssp, will turn on stack smashing
     protection (if available by compiler) by adding to CFLAGS.

   + Host, user, and pass fields have been increased in size.

   + New debugTimestamping boolean field has been added, which will
     cause the date and time to be printed with each call to the
     logging functions.


3.2.2.2, 2008-10-29

   + Large File Support has been updated internally to use newer APIs,
     rather than the transitional API.

   + You can now specify an IP address for binding the local end of
     sockets.  Useful if you need connections to originate from a
     specific IP on the system.

   + Bug fixed where binding an ephemeral port number in a specified
     range (rare) may not have used all available retries for binding.

   + Local writes now "retry" when the write returns less than the full
     number of bytes specified, and will call write() again to write
     out the remaining bytes.  This bug is very rare since local write
     almost always write out the full number of bytes.  This bug may
     have been limited to a particular type of filesystem.

   + Using a higher resolution timer on Windows, so that the timer
     no longer wraps over after a month.

   + The minimum version of Windows that the source code will now
     compile on is Windows 2000, although with a little tweaking you
     could compile for Windows NT 4 again.

   + Compatibility fix for Mac OS X when compiling from source to build
     universal binaries (Thanks, Toshi NAGATA)

   + Compatibility fix for FreeBSD, where a connection attempt may
     have failed with "Operation now in progress."


3.2.2, 2008-08-08

  + PASV mode will now be retried, in case of a transient failure.
    (Thanks, Steven Frank).

  + The directory listing parser now can recognize dates in the format
    of YYYY-mm-dd HH:MM[:SS].

  + Fixed feature detection for WS_FTP servers.

  + Workaround bug in REST on WS_FTP 6.0 (Thanks, Dan Nelson).

  + Fixed a problem where doing a get on a remote file that did not exist,
    but a file by the same name existed locally, would err out but truncate
    the local file anyway.

  + On Mac OS X, the configure script will now try to build universal
    binaries by default.  If you don't want that, pass --disable-universal
    to configure.

  + The default connect timeout has been reduced from 30 to 10 seconds to 
    reflect the increase in the typical user's network speed.

  + Visual Studio 2008 solution files are now provided for building the
    source on Windows.


3.2.1, 2007-07-29

  + Now a little more tolerant of FTP servers who send their internal network
    IP addresses in their PASV response.  When this is detected, it will be
    ignored and the data connection will connect to the same IP that is used
    for the control connection.

  + Fix problem where an empty directory could cause a recursive download
    to abort prematurely.
  
  + Resuming of uploads can now work for servers that do not support
    REST + STOR.  The APPE command will be used instead.

  + A few memory leaks plugged.

  + Fixes for FreeBSD 6 for compatibility with /usr/bin/make.


3.2.0, 2006-08-05

  + Fix read overflow in FTW (Thanks, Carl MAILLOUX, Craig Ball, and others).


libncftp/samples/ncftpget/ncftpget.c
/* ncftpget.c
 *
 * A non-interactive utility to grab files from a remote FTP server.
 * Very useful in shell scripts!
 *
 * NOTE:  This is for demonstration only -- an up-to-date version of this
 * program comes with the NcFTP Client 3.0 distribution.  This version
 * will most likely have a subset of features and fixes.
 */

#define VERSION "1.6.0"

#if (defined(WIN32) || defined(_WINDOWS)) && !defined(__CYGWIN__)
#	define _CRT_SECURE_NO_WARNINGS 1
#endif

#ifdef HAVE_CONFIG_H
#	include <config.h>
#endif

#include <ncftp.h>				/* Library header. */
#include <Strn.h>				/* Library header. */

#include "gpshare.h"

FTPLibraryInfo li;
FTPConnectionInfo fi;

extern int gFirewallType;
extern char gFirewallHost[64];
extern char gFirewallUser[32];
extern char gFirewallPass[32];
extern unsigned int gFirewallPort;

static void
Usage(void)
{
	FILE *fp;

	fp = OpenPager();
	(void) fprintf(fp, "NcFTPGet %s.\n\n", VERSION);
	(void) fprintf(fp, "*** NOTE: This version is for demonstration only -- get the latest and most\nfeature complete version from the NcFTP Client package. ***\n\n");
	(void) fprintf(fp, "Usages:\n");
	(void) fprintf(fp, "  ncftpget [flags] remote-host local-dir remote-path-names...   (mode 1)\n");
	(void) fprintf(fp, "  ncftpget -f login.cfg [flags] local-dir remote-path-names...  (mode 2)\n");
	(void) fprintf(fp, "  ncftpget [flags] ftp://url.style.host/path/name               (mode 3)\n");
	(void) fprintf(fp, "\nFlags:\n\
  -u XX  Use username XX instead of anonymous.\n\
  -p XX  Use password XX with the username.\n\
  -P XX  Use port number XX instead of the default FTP service port (21).\n\
  -d XX  Use the file XX for debug logging.\n\
  -a     Use ASCII transfer type instead of binary.\n");
	(void) fprintf(fp, "\
  -t XX  Timeout after XX seconds.\n\
  -v/-V  Do (do not) use progress meters.\n\
  -f XX  Read the file XX for host, user, and password information.\n\
  -A     Append to local files, instead of overwriting them.\n");
	(void) fprintf(fp, "\
  -z/-Z  Do (do not) not try to resume downloads (default: -z).\n\
  -F     Use passive (PASV) data connections.\n\
  -DD    Delete remote file after successfully downloading it.\n\
  -r XX  Redial XX times until connected.\n\
  -R     Recursive mode; copy whole directory trees.\n");
	(void) fprintf(fp, "\nExamples:\n\
  ncftpget ftp.wustl.edu . /pub/README /pub/README.too\n\
  ncftpget ftp.wustl.edu . '/pub/README*'\n\
  ncftpget -R ftp.ncftp.com /tmp /pub/ncftpd  (ncftpd is a directory)\n\
  ncftpget ftp://ftp.wustl.edu/pub/README\n\
  ncftpget -u gleason -p my.password Bozo.probe.net . '/home/mjg/.*rc'\n\
  ncftpget -u gleason Bozo.probe.net . /home/mjg/foo.txt  (prompt for password)\n\
  ncftpget -f Bozo.cfg '/home/mjg/.*rc'\n\
  ncftpget -a -d /tmp/debug.log -t 60 ftp.wustl.edu . '/pub/README*'\n");

	(void) fprintf(fp, "\nLibrary version: %s.\n", gLibNcFTPVersion + 5);
	(void) fprintf(fp, "\nThis is a freeware program by Mike Gleason (http://www.NcFTP.com/contact/).\n");
	(void) fprintf(fp, "This was built using LibNcFTP (http://www.ncftp.com/libncftp).\n");

	ClosePager(fp);
	DisposeWinsock();
	exit(kExitUsage);
}	/* Usage */





static void
Abort(int sigNum)
{
	signal(sigNum, Abort);

	/* Hopefully the I/O operation in progress
	 * will complete, and we'll abort before
	 * it starts a new block.
	 */
	fi.cancelXfer++;

	/* If the user appears to be getting impatient,
	 * restore the default signal handler so the
	 * next ^C abends the program.
	 */
	if (fi.cancelXfer >= 2)
		signal(sigNum, SIG_DFL);
}	/* Abort */




static int 
Copy(FTPCIPtr cip, const char *dstdir, char **files, int rflag, int xtype, int resumeflag, int appendflag, int deleteflag, int tarflag)
{
	int i;
	int result;
	const char *file;
	int rc = 0;

	for (i=0; ; i++) {
		file = files[i];
		if (file == NULL)
			break;
		result = FTPGetFiles3(cip, file, dstdir, rflag, kGlobYes, xtype, resumeflag, appendflag, deleteflag, tarflag, kNoFTPConfirmResumeDownloadProc, 0);
		if (result != 0) {
			(void) fprintf(stderr, "ncftpget: file retrieval error: %s.\n", FTPStrError(result));
			rc = result;
		}
	}
	return (rc);
}	/* Copy */



main_void_return_t
main(int argc, char **argv)
{
	int result, c;
	int rflag = 0;
	int xtype = kTypeBinary;
	int appendflag = kAppendNo;
	int resumeflag = kResumeYes;
	int deleteflag = kDeleteNo;
	int tarflag = kTarYes;
	int progmeters;
	const char *dstdir;
	char **flist;
	ExitStatus es;
	char url[256];
	char urlfile[128];
	int urlxtype;
	FTPLineList cdlist;
	FTPLinePtr lp;
	int rc;
	int nD = 0;
	GetoptInfo opt;

	InitWinsock();
	result = FTPInitLibrary(&li);
	if (result < 0) {
		(void) fprintf(stderr, "ncftpget: init library error %d (%s).\n", result, FTPStrError(result));
		DisposeWinsock();
		exit(kExitInitLibraryFailed);
	}
	result = FTPInitConnectionInfo(&li, &fi, kDefaultFTPBufSize);
	if (result < 0) {
		(void) fprintf(stderr, "ncftpget: init connection info error %d (%s).\n", result, FTPStrError(result));
		DisposeWinsock();
		exit(kExitInitConnInfoFailed);
	}

	fi.debugLog = NULL;
	fi.errLog = stderr;
	fi.xferTimeout = 60 * 60;
	fi.connTimeout = 30;
	fi.ctrlTimeout = 135;
	(void) STRNCPY(fi.user, "anonymous");
	fi.host[0] = '\0';
	progmeters = GetDefaultProgressMeterSetting();
	urlfile[0] = '\0';
	InitLineList(&cdlist);
	dstdir = NULL;

	GetoptReset(&opt);
	while ((c = Getopt(&opt, argc, argv, "P:u:p:e:d:t:aRr:vVf:ADzZFT")) > 0) switch(c) {
		case 'P':
			fi.port = atoi(opt.arg);	
			break;
		case 'u':
			(void) STRNCPY(fi.user, opt.arg);
			break;
		case 'p':
			(void) STRNCPY(fi.pass, opt.arg);	/* Don't recommend doing this! */
			break;
		case 'e':
			if (strcmp(opt.arg, "stdout") == 0)
				fi.errLog = stdout;
			else if (opt.arg[0] == '-')
				fi.errLog = stdout;
			else if (strcmp(opt.arg, "stderr") == 0)
				fi.errLog = stderr;
			else
				fi.errLog = fopen(opt.arg, "a");
			break;
		case 'D':
			/* Require two -D's in case they typo. */
			nD++;
			break;
		case 'd':
			if (strcmp(opt.arg, "stdout") == 0)
				fi.debugLog = stdout;
			else if (opt.arg[0] == '-')
				fi.debugLog = stdout;
			else if (strcmp(opt.arg, "stderr") == 0)
				fi.debugLog = stderr;
			else
				fi.debugLog = fopen(opt.arg, "a");
			break;
		case 't':
			SetTimeouts(&fi, opt.arg);
			break;
		case 'a':
			xtype = kTypeAscii;
			break;
		case 'r':
			SetRedial(&fi, opt.arg);
			break;
		case 'R':
			rflag = 1;
			break;
		case 'T':
			tarflag = 0;
			break;
		case 'v':
			progmeters = 1;
			break;
		case 'V':
			progmeters = 0;
			break;
		case 'f':
			ReadConfigFile(opt.arg, &fi);
			break;
		case 'A':
			appendflag = kAppendYes;
			break;
		case 'z':
			resumeflag = kResumeYes;
			break;
		case 'Z':
			resumeflag = kResumeNo;
			break;
		case 'F':
			if (fi.dataPortMode == kPassiveMode)
				fi.dataPortMode = kSendPortMode;
			else
				fi.dataPortMode = kPassiveMode;
			break;
		default:
			Usage();
	}
	if (opt.ind > argc - 1)
		Usage();

	InitOurDirectory();
	LoadFirewallPrefs();

	if (progmeters != 0)
		fi.progress = PrStatBar;

	if (fi.host[0] == '\0') {
		(void) STRNCPY(url, argv[opt.ind]);
		rc = FTPDecodeURL(&fi, url, &cdlist, urlfile, sizeof(urlfile), (int *) &urlxtype, NULL);
		if (rc == kMalformedURL) {
			(void) fprintf(stderr, "Malformed URL: %s\n", url);
			DisposeWinsock();
			exit(kExitMalformedURL);
		} else if (rc == kNotURL) {
			/* This is what should happen most of the time. */
			if (opt.ind > argc - 3)
				Usage();
			(void) STRNCPY(fi.host, argv[opt.ind]);
			dstdir = argv[opt.ind + 1];
			flist = argv + opt.ind + 2;
		} else {
			/* URL okay */
			flist = NULL;
			if ((urlfile[0] == '\0') && (rflag == 0)) {
				/* It was obviously a directory, and they didn't say -R. */
				(void) fprintf(stderr, "ncftpget: Use -R if you want the whole directory tree.\n");
				es = kExitUsage;
				DisposeWinsock();
				exit((int) es);
			}
			xtype = urlxtype;
		}
	} else {
		if (opt.ind > argc - 2)
			Usage();
		dstdir = argv[opt.ind + 0];
		flist = argv + opt.ind + 1;
	}

	if ((strcmp(fi.user, "anonymous") != 0) && (strcmp(fi.user, "ftp") != 0) && (fi.pass[0] == '\0'))
		(void) GetPass("Password: ", fi.pass, sizeof(fi.pass));

	if (MayUseFirewall(fi.host) != 0) {
		fi.firewallType = gFirewallType; 
		(void) STRNCPY(fi.firewallHost, gFirewallHost);
		(void) STRNCPY(fi.firewallUser, gFirewallUser);
		(void) STRNCPY(fi.firewallPass, gFirewallPass);
		fi.firewallPort = gFirewallPort;
	}

	if (nD >= 2)
		deleteflag = kDeleteYes;
	
	es = kExitOpenTimedOut;
	if ((result = FTPOpenHost(&fi)) < 0) {
		(void) fprintf(stderr, "ncftpget: cannot open %s: %s.\n", fi.host, FTPStrError(result));
		es = kExitOpenFailed;
		DisposeWinsock();
		exit((int) es);
	}
	if (flist == NULL) {
		/* URL mode */
		es = kExitChdirTimedOut;
		for (lp = cdlist.first; lp != NULL; lp = lp->next) {
			if (FTPChdir(&fi, lp->line) != 0) {
				(void) fprintf(stderr, "ncftpget: cannot chdir to %s: %s.\n", lp->line, FTPStrError(fi.errNo));
				es = kExitChdirFailed;
				DisposeWinsock();
				exit((int) es);
			}
		}
		
		es = kExitXferTimedOut;
		(void) signal(SIGINT, Abort);
		if (FTPGetFiles3(&fi, urlfile, ".", rflag, kGlobYes, xtype, resumeflag, appendflag, deleteflag, tarflag, kNoFTPConfirmResumeDownloadProc, 0) < 0) {
			(void) fprintf(stderr, "ncftpget: file retrieval error: %s.\n", FTPStrError(fi.errNo));
			es = kExitXferFailed;
		} else {
			es = kExitSuccess;
		}
	} else {
		es = kExitXferTimedOut;
		(void) signal(SIGINT, Abort);
		if (Copy(&fi, dstdir, flist, rflag, xtype, resumeflag, appendflag, deleteflag, tarflag) < 0)
			es = kExitXferFailed;
		else
			es = kExitSuccess;
	}
	(void) FTPCloseHost(&fi);
	DisposeWinsock();
	exit((int) es);
}	/* main */
